### ISSUE TITLE
Categorical data not surviving a repartition

### ISSUE THREAD

**Issue by shughes-uk** [dask/dask#2674]
When creating categorical columns with  
```
df[column] = df[column].astype('category')
df = df.repartition(npartitions=n)
```
  seems to result in partitions that have the old datatype for the column and one partition that has the correct category datatype.

This is with dask 15.2, I have not tested older versions.

**Comment by mrocklin** [comment]
This boils down to `pd.concat` not propagating categories when they don't match

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': ['a', 'b', 'a', 'c']})

In [3]: df2 = pd.DataFrame({'x': [1, 2, 3, 4], 'y': ['a', 'b', 'a', 'e']})

In [4]: df['y'] = df['y'].astype('category')

In [5]: df2['y'] = df2['y'].astype('category')

In [6]: df.dtypes
Out[6]: 
x       int64
y    category
dtype: object

In [7]: df2.dtypes
Out[7]: 
x       int64
y    category
dtype: object

In [8]: pd.concat([df, df2]).dtypes
Out[8]: 
x     int64
y    object
dtype: object
```

@TomAugspurger is there a way around this?

(my test case for future use)

```python
@pytest.mark.parametrize('npartitions', [1, 4])
def test_repartition_on_categoricals(npartitions):
    df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': ['a', 'b', 'a', 'c']})
    ddf = dd.from_pandas(df, npartitions=2)
    ddf['y'] = ddf['y'].astype('category')
    ddf2 = ddf.repartition(npartitions=npartitions)
    
    df = df.copy() 
    df['y'] = df['y'].astype('category')
    assert_eq(df, ddf)
    assert_eq(df, ddf2)
```

**Comment by martindurant** [comment]
Isn't that surprising from `repartition`, where the partitions *were* known beforehand and do not get recoded?

**Comment by jcrist** [comment]
This might be fixed by replacing the calls to `pd.concat` with `dask.dataframe.methods.concat` (which handles mixed categories).

**Comment by jcrist** [comment]
Yep, PR coming momentarily.

**Comment by mrocklin** [comment]
Thank you @jcrist

**Comment by TomAugspurger** [comment]
For reference, here's the pandas issue about concat and categoricals https://github.com/pandas-dev/pandas/issues/14177 (we weren't sure whether unioning, converting to object, or raising was "correct").

**Comment by jcrist** [comment]
Should be fixed by #2676.
